src/natcap/invest/habitat_quality.py

```diff
@@ -187,282 +190,335 @@
-
-            # Remove the nodata value from the set of landuser codes.
-            nodata = pygeoprocessing.get_raster_info(lulc_path)['nodata'][0]
-            try:
-                raster_unique_lucodes.remove(nodata)
-            except KeyError:
-                # KeyError when the nodata value was not encountered in the
-                # raster's pixel values.  Same result when nodata value is
-                # None.
-                pass
+            unique_lucode_task = task_graph.add_task(
+                func=_collect_unique_lucodes,
+                args=((lulc_path, 1), ),
+                task_name=f'unique_lucodes{lulc_key}')
+            unique_lucode_task_list.append(unique_lucode_task)
 
             # add a key to the threat dictionary that associates all threat
             # rasters with this land cover
             threat_path_dict['threat' + lulc_key] = {}
 
             # for each threat given in the CSV file try opening the associated
-            # raster which should be found in threat_raster_folder
+            # raster which should be found relative to the Threat CSV
             for threat in threat_dict:
-                # it's okay to have no threat raster for baseline scenario
-                threat_path_dict['threat' + lulc_key][threat] = (
-                    resolve_ambiguous_raster_path(
-                        os.path.join(threat_raster_dir, threat + lulc_key),
-                        raise_error=(lulc_key != '_b')))
-
+                LOGGER.debug(f"Validating path for threat: {threat}")
+                # Build absolute threat path from threat table
+                threat_table_path_col = _THREAT_SCENARIO_MAP[lulc_key]
+                threat_path_relative = (
+                    threat_dict[threat][threat_table_path_col])
+                threat_path = os.path.join(
+                    threat_csv_dirpath, threat_path_relative)
+
+                # check gis type of threat path and catch thrown ValueError
+                # from get_gis_type if path does not exist
+                try:
+                    threat_gis_type = pygeoprocessing.get_gis_type(threat_path)
+                    # if threat path not of type RASTER then raise value error
+                    if threat_gis_type != pygeoprocessing.RASTER_TYPE:
+                        raise ValueError
+                except ValueError:
+                    # it's okay to have no threat raster for baseline scenario
+                    if lulc_key != '_b':
+                        raise ValueError(
+                            'There was an Error locating a threat raster from '
+                            'the path in CSV for column: '
+                            f'{_THREAT_SCENARIO_MAP[lulc_key]} and threat: '
+                            f'{threat}. The path in the CSV column should be '
+                            'relative to the threat CSV table.')
+                    else:
+                        threat_path = None
```

Comment:
I refactored this here: https://github.com/natcap/invest/pull/75/commits/36fd0d9aa4f02357bab8bb5d47c26b7b03d8998f

I broke it out into its own function since the same code was being used twice, once in the model and once for validation.

Reply:
Great!